### PR TITLE
C++ : NULL application name with an unquoted path in call to CreateProcess

### PR DESCRIPTION
**Pull request by raulgarciamsft**
Calling a function of the CreatePorcess* family of functions, which may result in a security vulnerability if the path contains spaces.

### CODE REVIEW

File: cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.ql [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+      val = sink.asExpr() |
+      val = call.getArgument(call.getApplicationNameArgumentId()) 
+    )
+  }
+}
+
+/**
+ * Dataflow that detects a call to CreateProcess with an unquoted commandLine argument
+ */
+class QuotedCommandInCreateProcessFunctionConfiguration extends DataFlow2::Configuration {
```
Comment (geoffw0) [inline cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.ql]: I think this class does the opposite of what it's name implies.  `UnquotedCommandInCreateProcessFunctionConfiguration`?

File: cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.ql [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+      this.getTarget().hasGlobalName("CreateProcessWithTokenW") 
+    ) then ( result = 2 )
+    else if (
+      this.getTarget().hasGlobalName("CreateProcessWithLogonW") 
+    ) then ( result = 4 )
+    else if(
+      this.getTarget().hasGlobalName("CreateProcessAsUserA") or
+      this.getTarget().hasGlobalName("CreateProcessAsUserW") 
+    ) then ( result = 1 )
+    else (result = -1 )
```
Comment (geoffw0) [inline cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.ql]: The `-1` cases in this predicate and the one below could be omitted.  It's perfectly fine for a predicate to have no result, and either way a use of `FunctionCall.getArgument(thisPredicate())` would have no result.

File: cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.ql [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+ * Dataflow that detects a call to CreateProcess with an unquoted commandLine argument
+ */
+class QuotedCommandInCreateProcessFunctionConfiguration extends DataFlow2::Configuration {
+  QuotedCommandInCreateProcessFunctionConfiguration() {
+    this = "QuotedCommandInCreateProcessFunctionConfiguration"
+  }
+
+  override predicate isSource(DataFlow2::Node source) {
+    exists( string s |
+      s = source.asExpr().getValue().toString()
```
Comment (geoffw0) [inline cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.ql]: At this point we could determine whether the string contains any spaces.  Is there any reason we should report an unquoted string if it doesn't?

File: cpp/ql/test/query-tests/Security/CWE/CWE-428/UnsafeCreateProcessCall.cpp [hunk 4]
```diff
@@ -0,0 +1,503 @@
+// semmle-extractor-options: --microsoft
+#define NULL 0
+#define FALSE 0
+#define far
+#define LOGON_WITH_PROFILE 0x00000001
+
+typedef char CHAR;
```
Comment (dave-bartolomeo) [inline cpp/ql/test/query-tests/Security/CWE/CWE-428/UnsafeCreateProcessCall.cpp]: Would it be possible to remove most of these typedefs, and leave the various struct definitions with empty bodies (e.g. `struct _STARTUPINFOW { };`? This shouldn't affect the query, and all of these cut-and-paste definitions from the Windows headers kind of clutter up the real test cases.

File: cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.ql [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+ */
+ 
+import cpp
+import semmle.code.cpp.dataflow.DataFlow
+import semmle.code.cpp.dataflow.DataFlow2
+
+/**
+ * A function call to CreateProcess (either wide-char or single byte string versions)
+ */
+class CreateProcessFunctionCall extends FunctionCall {
```
[the remark this answers is not in the document]
Comment (raulgarciamsft) [reply cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.ql]: I will fix it & submit the change

File: cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.ql (hunk 2 above)
Comment (raulgarciamsft) [reply cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.ql]: Got it. I will fix it & submit the change

File: cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.ql (hunk 1 above)
Comment (raulgarciamsft) [reply cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.ql]: Good Catch. I started writing it the other way around, and I forgot to change the name. I will fix it

File: cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.ql (hunk 3 above)
Comment (raulgarciamsft) [reply cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.ql]: I originally thought doing on the same predicate I test for quotes, but I wasn't sure if it is worth just testing if the string has spaces/tabs or if I should add other checks (i.e. using environment variables, other interesting characters/sequences). If you think it is worth just adding a simple check for space/tab, I can add it.

**Comment by raulgarciamsft** [comment]
BTW. After talking with Robert & Dave offline. I will try to make some major changes. Thanks

File: cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.ql (hunk 2 above)
Comment (rdmarsh2) [reply cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.ql]: The best thing to do is to rewrite those predicates using `and` and `or`:
```
(
  this.getTarget().hasGlobalName("CreateProcessA") or
  this.getTarget().hasGlobalName("CreateProcessW")
) and
result = 0
or
...
```

Comment (raulgarciamsft) [reply cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.ql]: Never mind, with Dave's recommendation below, I don't need the IF anymore. Thanks

File: cpp/ql/test/query-tests/Security/CWE/CWE-428/UnsafeCreateProcessCall.cpp (hunk 4 above)
Comment (raulgarciamsft) [reply cpp/ql/test/query-tests/Security/CWE/CWE-428/UnsafeCreateProcessCall.cpp]: No problem. I have been trying to make the test function calls as close to the real ones, but I will change it.

File: cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.ql (hunk 3 above)
Comment (raulgarciamsft) [reply cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.ql]: Fixed.

File: cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.ql (hunk 5 above)
[the remark this answers is not in the document]
Comment (raulgarciamsft) [reply cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.ql]: Fixed

File: cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.qhelp [hunk 6]
```diff
@@ -0,0 +1,46 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+
+<overview>
+  <p>This query indicates that there is a call to a function of the <code>CreateProcess*</code> family of functions, which may result in a security vulnerability if the path contains spaces.</p>
```
Comment (semmledocs-ac) [inline cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.qhelp]: ```which may result in a security vulnerability if the path contains spaces```
Perhaps this is a little tentative? Perhaps we should say that the code identified by the query _does_ introduce a vulnerability. Change "which may result in a security vulnerability" to:
```which introduces a security vulnerability```

File: cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.ql [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    appName = call.getArgument(call.getApplicationNameArgumentId())
+    and nullAppConfig.hasFlow(DataFlow2::exprNode(source), DataFlow2::exprNode(appName))
+    and msg1 = call.toString() + " with lpApplicationName == NULL (" + appName + ")"
+  )
+  and
+  exists( Expr source, Expr cmd,
+    QuotedCommandInCreateProcessFunctionConfiguration quotedConfig |
+    cmd = call.getArgument(call.getCommandLineArgumentId())
+    and quotedConfig.hasFlow(DataFlow2::exprNode(source), DataFlow2::exprNode(cmd))
+    and msg2 = " and with an unquoted lpCommandLine (" + cmd + ") may result in a security vulnerability if the path contains spaces."
```
Comment (semmledocs-ac) [inline cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.ql]: As per comment above, consider changing 
```may result in a security vulnerability```
to 
```introduces a security vulnerability```

**Comment by raulgarciamsft** [comment]
I should have pasted the offline conversation summary here, sorry about that:

We were trying to discuss if there was a way to detect this scenario when a developer uses a writable buffer (as the API intends) instead of doing an unsafe cast from a static read-only string to a writable string, but we cannot easily check if the unquoted string is copied ( i.e.  lpComamndLine = new wchar_t[256]; wcscpy_s(lpComamndLine, 256, L"c:\\program files\\foo"); )

This version of the Semmle rule (with the changes recommended on the PR) can detect most of the cases with a static string, improving over the original PreFAST rule that misses anything that is not hardcoded directly on the CreateProcess* function call, and that adding the check for the lpCommandLine argument to have a space/tab helps to avoid a false positive.

Historically the check for C6277 made sense for a time when the majority of the Windows apps were not using Unicode, and the CreateProcess*A versions of the API would work with a string literal (circa Windows 9x), so we are not expecting a lot of hits on this one.

File: cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.qhelp (hunk 6 above)
Comment (raulgarciamsft) [reply cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.qhelp]: Fixed

File: cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.qhelp [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+</overview>
+
+<recommendation>
+<p>Do not use <code>NULL</code> for the <code>lpApplicationName</code> argument to the <code>CreateProcess*</code> function.</p>
+<p>If you pass <code>NULL</code> for <code>lpApplicationName</code>, use quotation marks around the executable path in <code>lpCommandLine</code>.</p>
+</recommendation>
+
+<example>
+<p>In the following example, <code>CreateProcessW</code> is called with a NULL value for <code>lpApplicationName</code>,
```
[the remark this answers is not in the document]
Comment (raulgarciamsft) [reply cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.qhelp]: Fixed

File: cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.qhelp [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+<recommendation>
+<p>Do not use <code>NULL</code> for the <code>lpApplicationName</code> argument to the <code>CreateProcess*</code> function.</p>
+<p>If you pass <code>NULL</code> for <code>lpApplicationName</code>, use quotation marks around the executable path in <code>lpCommandLine</code>.</p>
+</recommendation>
+
+<example>
+<p>In the following example, <code>CreateProcessW</code> is called with a NULL value for <code>lpApplicationName</code>, 
+and the value for <code>lpCommandLine</code> that represent the application path is not quoted and has spaces in it.</p>
+<p>If an attacker has access to the file system, it is possible to elevate privileges by creating a file such as "C:\Program.exe" that will be executed instead of the intended application.</p>
```
[the remark this answers is not in the document]
Comment (raulgarciamsft) [reply cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.qhelp]: Fixed

File: cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.ql [hunk 10]
```diff
@@ -0,0 +1,129 @@
+/**
+ * @name NULL application name with an unquoted path in call to CreateProcess
+ * @description Calling a function of the CreateProcess* family of functions, which may result in a security vulnerability if the path contains spaces.
```
[the remark this answers is not in the document]
Comment (raulgarciamsft) [reply cpp/ql/src/Security/CWE/CWE-428/UnsafeCreateProcessCall.ql]: Fixed

**Comment by semmledocs-ac** [comment]
@raulgarciams - thanks for making those changes

**Comment by semmledocs-ac** [comment]
@raulgarciamsft  - thanks for making those changes

**Comment by geoffw0** [comment]
Great.  Support for string buffers sounds like a desirable improvement but I'm very happy with this as a solid initial version of the query.  I'll merge this in a few hours unless anybody speaks up (@dave-bartolomeo ?).

**Pull request merged**